src/iidm/util/TerminalFinder.cpp

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright (c) 2020, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <powsybl/iidm/util/TerminalFinder.hpp>
+
+#include <powsybl/iidm/Branch.hpp>
+#include <powsybl/iidm/BusbarSection.hpp>
+#include <powsybl/iidm/HvdcConverterStation.hpp>
+#include <powsybl/iidm/Injection.hpp>
+#include <powsybl/iidm/TwoWindingsTransformer.hpp>
+#include <powsybl/stdcxx/instanceof.hpp>
+
+namespace powsybl {
+
+namespace iidm {
+
+namespace util {
+
+TerminalFinder::TerminalFinder(const std::vector<Predicate>& predicates) :
+    m_comparator(getComparator(predicates)) {
+}
+
+stdcxx::Reference<Terminal> TerminalFinder::find(const stdcxx::range<Terminal>& terminals) const {
+    auto it = std::max_element(terminals.begin(), terminals.end(), m_comparator);
+    return it != terminals.end() ? stdcxx::ref(*it) : stdcxx::ref<Terminal>();
+}
+
+TerminalFinder::Comparator TerminalFinder::getComparator(const std::vector<Predicate>& predicates) {
+    return [&predicates](Terminal& t1, Terminal& t2) {
+        if (stdcxx::areSame(t1, t2)) {
+            return false;
+        }
+        for (const auto& predicate : predicates) {
+            if (predicate(t1) || predicate(t2)) {
+                return !predicate(t1) && predicate(t2);
+            }
+        }
+        return false; // no predicate is verified by t1 or t2, the first is then considered better than the latter
+    };
+}
+
+TerminalFinder TerminalFinder::getDefault() {
+    return TerminalFinder(getDefaultRules());
+}
+
+const std::vector<TerminalFinder::Predicate>& TerminalFinder::getDefaultRules() {
+    static std::vector<Predicate> s_rules = {
+        [](Terminal& t) { return t.getConnectable() && stdcxx::isInstanceOf<BusbarSection>(t.getConnectable().get()); },
+        [](Terminal& t) { return t.getConnectable() && stdcxx::isInstanceOf<Injection>(t.getConnectable().get()); },
+        [](Terminal& t) { return t.getConnectable() && stdcxx::isInstanceOf<Branch>(t.getConnectable().get()); },
+        [](Terminal& t) { return t.getConnectable() && stdcxx::isInstanceOf<ThreeWindingsTransformer>(t.getConnectable().get()); },
+        [](Terminal& t) { return t.getConnectable() && stdcxx::isInstanceOf<HvdcConverterStation>(t.getConnectable().get()); }
```

Comment:
It would be nice to create a new version of the `stdcxx::isInstanceOf` that takes a reference_wrapper<T>.